### PR TITLE
Experimental amazon lookup API and existing API refactor

### LINKED ISSUE #2334: Amazon API results give overspecific (made up) day and month for older books
### Is your feature request related to a problem? Please describe.

Older books sourced from Amazon metadata may have over specific dates from earlier than a particular year.

Needs investigation, but Amazon sourced dates have a particular format, with day, month, year, Beyond a certain point, the API seems to use the current day and month because their record does not actually have one -- I'll need to confirm this, and check how much of our exisiting data is affected, but I wanted to create the issue to remind myself to investigate further.

It means the publication year is good, but month and day portions are likely junk.

3 parts:

- [x] Investigate and confirm the issue, id the period where specific dates are good vs. bad
- [ ] Identify existing records affected (matching criteria to locate them in dump files and count of affected records)
- [ ] Adjust import code to process these dates appropriately
- [ ] Clean up existing records

### PR DESCRIPTION
**Pull request by hornc**
### Description

* Adds an admin / librarian only endpoint that can be used to look up book prices by title and author (for investigating what data is provided from the Amazon search API)
* Refactors the existing price lookup API to make this possible (split some code out to a new `_serialize_amazon_product()` method) and structure the code in a more usable fashion. This is potentially going to help with #2332, and unifying our data source usage
* Minor improvements around date handling, results not found exception handling, and API response format, which will also apply to the standard price lookup results, which does affect every Edition page for books with ISBN.

I did this work as part of a data investigation task, and it seems a shame to throw away the code. The refactor is the main useful aspect of this PR, although the debugging and extra insight into the capabilities of the Amazon API has potential to be useful later.

Does _not_ fully fix #2334 (Import data format issues), although it has helped gather data on the problem.  I will raise a separate PR once this is merged. This sort of librarian tool will help trouble-shoot similar issues, or develop new features, in future.

Closes #

### Technical

The current endpoint is at `/ext?title=foo&author=bar`  It should have a better name. 'ext' is short for 'external search', so perhaps `/external_search?title=foo&author=bar` is more appropriate.

### CODE REVIEW

**Comment by hornc** [comment]
@cdrini description updated, any suggestions for a better url path and naming would be much appreciated :) Currently it is librarian / admin only by design.

**Comment by hornc** [comment]
@mekarpeles any opinions on what this API url should be?

**Comment by cdrini** [comment]
Ah, actually I can't test this cause I don't have access to dev; so un-assigning myself.

File: openlibrary/core/vendors.py [hunk 1]
```diff
@@ -21,24 +22,19 @@ def get_amazon_metadata(id_, id_type='isbn'):
...
-    kwargs = {}
-    if id_type == 'isbn':
-        id_ = normalize_isbn(id_)
-        kwargs = {'SearchIndex': 'Books', 'IdType': 'ISBN'}
-    kwargs['ItemId'] = id_
+def get_amazon_search(title='', author=''):
+    # uncached for now, cache later
+    kwargs = {'Title': title, 'Author': author, 'SearchIndex': 'Books'}
+    results = lending.amazon_api.search(**kwargs)
+    data = []
```
Comment (hornc) [inline openlibrary/core/vendors.py]: this should be a dict  `{"results": []}` to be converted into JSON

File: openlibrary/core/vendors.py [hunk 2]
```diff
@@ -21,24 +22,19 @@ def get_amazon_metadata(id_, id_type='isbn'):
...
-            raise Exception
-        product = lending.amazon_api.lookup(**kwargs)
-        # sometimes more than one product can be returned, choose first
-        if isinstance(product, list):
-            product = product[0]
-    except Exception as e:
-        return None
+        for product in results:
+            data.append(_serialize_amazon_product(product))
+    except SearchException:
```
[the remark this answers is not in the document]
Comment (hornc) [reply openlibrary/core/vendors.py]: `results` is an iterator at this point, the api call is actually occurring in the `try` block https://github.com/yoavaviram/python-amazon-simple-product-api/blob/master/amazon/api.py#L267

File: openlibrary/plugins/openlibrary/api.py [hunk 3]
```diff
@@ -242,12 +242,27 @@ def get_works_data(self, author, limit, offset):
             "entries": works
         }
 
+class external_search_api(delegate.page):
+    path = '/ext'
```
[the remark this answers is not in the document]
Comment (hornc) [reply openlibrary/plugins/openlibrary/api.py]: actually, how about `/_tools/amazon_search` ?  this uses the Amazon termed 'ItemSearch' API https://docs.aws.amazon.com/AWSECommerceService/latest/DG/ItemSearch.html, while our `/isbn` endpoint uses the 'ItemLookup' API endpoint, using the Amazon terms.

File: openlibrary/plugins/openlibrary/api.py [hunk 4]
```diff
@@ -242,12 +242,27 @@ def get_works_data(self, author, limit, offset):
             "entries": works
         }
 
+class external_search_api(delegate.page):
```
[the remark this answers is not in the document]
Comment (hornc) [reply openlibrary/plugins/openlibrary/api.py]: renaming to `amazon_search_api`, because there is a separate Amazon lookup API, and this isn't it :)

File: openlibrary/core/vendors.py [hunk 5]
```diff
@@ -167,7 +246,32 @@ def cached_get_amazon_metadata(*args, **kwargs):
...
+    :return: Metadata for a single BWB book, currently listed on their catalog, or error dict.
+    :rtype: dict
+    """
+
+    isbn = normalize_isbn(isbn)
+    try:
+        if isbn:
+            return _get_betterworldbooks_metadata(isbn)
+    except Exception:
+        return {}
```
Comment (hornc) [inline openlibrary/core/vendors.py]: The return values here need testing, it looks like `None` is a possible return value (because `normalize_isbn()` _can_ return `None` if the value is not isbn like at all) , but I'm not sure `{}` can ever be returned as exceptions are handled in `_get_betterworldbooks_metadata()` and meaningful error json is returned. Possibly for another PR, this PR is meant to be Amazon focused, but I filled in docstrings for everything while I was on a roll.

Comment (cdrini) [reply openlibrary/core/vendors.py]: > but I filled in docstrings for everything while I was on a roll

😍 

> Possibly for another PR, this PR is meant to be Amazon focused

That seems reasonable to me.

**Comment by mekarpeles** [comment]
This is lgtm
